atlasdb-config/src/main/java/com/palantir/atlasdb/factory/BroadcastDialogueServiceProvider.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.factory;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Maps;
+import com.palantir.atlasdb.config.AuxiliaryRemotingParameters;
+import com.palantir.atlasdb.config.ImmutableAuxiliaryRemotingParameters;
+import com.palantir.atlasdb.config.ImmutableServerListConfig;
+import com.palantir.atlasdb.config.ServerListConfig;
+import com.palantir.atlasdb.http.AtlasDbHttpProtocolVersion;
+import com.palantir.atlasdb.http.v2.DialogueClientOptions;
+import com.palantir.atlasdb.http.v2.DialogueShimFactory;
+import com.palantir.atlasdb.http.v2.ImmutableRemoteServiceConfiguration;
+import com.palantir.atlasdb.http.v2.RemoteServiceConfiguration;
+import com.palantir.common.streams.KeyedStream;
+import com.palantir.conjure.java.api.config.service.UserAgent;
+import com.palantir.dialogue.Channel;
+import com.palantir.dialogue.clients.DialogueClients;
+import com.palantir.refreshable.Refreshable;
+
+/**
+ * Provides a mechanism for creating individual proxies that use Dialogue for communication for each node in the cluster.
+ *
+ * Furthermore, proxies should include in their {@link com.palantir.conjure.java.api.config.service.UserAgent}
+ * information to allow client services to identify the protocol they are using to talk, via
+ * {@link AtlasDbHttpProtocolVersion}.
+ */
+public final class BroadcastDialogueServiceProvider {
+    DialogueClients.ReloadingFactory reloadingFactory;
+    Refreshable<ServerListConfig> serverListConfigSupplier;
+
+    private BroadcastDialogueServiceProvider(DialogueClients.ReloadingFactory reloadingFactory,
+            Refreshable<ServerListConfig> serverListConfigSupplier) {
+        this.reloadingFactory = reloadingFactory;
+        this.serverListConfigSupplier = serverListConfigSupplier;
+    }
+
+    public static BroadcastDialogueServiceProvider create(
+            DialogueClients.ReloadingFactory baseFactory,
+            Refreshable<ServerListConfig> serverListConfigSupplier,
+            UserAgent versionedAgent,
```

Comment:
The passed in user-agent isn't the versioned one though: it needs to have the `atlasdb-http-client` added as an auxiliary agent (see AtlasDbDialogueServiceProvider for how this is done), so that we can respond correctly to 429s and 503s (even if we aren't going to get 308s, we need to know how to deal with the other two)